File: src/Matrix.ts
export class Matrix {
  M: number;
  N: number;

  private items: Float64Array;

  constructor(items: Float64Array, M: number, N: number) {
    this.M = M;
    this.N = N;
    this.items = items;
  }

  static fromArrays(items: number[][]) {
    const m = items.length;
    const n = items[0].length;

    const list = new Float64Array(m * n);
    for (let i = 0; i < m; i += 1) {
      for (let j = 0; j < n; j += 1) {
        list[i * n + j] = items[i][j];
      }
    }
    return new Matrix(list, m, n);
  }

  static fromList(items: number[], isVertical = true) {
    const m = isVertical ? items.length : 1;
    const n = isVertical ? 1 : items.length;

    const list = new Float64Array(items.length);
    for (let i = 0; i < items.length; i += 1) {
      list[i] = items[i];
    }

    return new Matrix(list, m, n);
  }

  get(i: number, j: number) {
    return this.items[calcIndex(i, j, this.N)];
  }

  set(i: number, j: number, value: number) {
    this.items[calcIndex(i, j, this.N)] = value;
  }

  toArrays() {
    const result: number[][] = [];
    for (let i = 0; i < this.M; i += 1) {
      result[i] = [];
      for (let j = 0; j < this.N; j += 1) {
        result[i][j] = this.get(i, j);
      }
    }
    return result;
  }

  multiply(matrix: Matrix): Matrix {
    if (this.N !== matrix.M)
      throw new Error(
        `Dimension mismatch. Got ${this.M}x${this.N} and ${matrix.M}x${matrix.N}`
      );
    const result = new Float64Array(this.M * matrix.N);
    for (let i = 0; i < this.M; i++) {
      for (let j = 0; j < matrix.N; j++) {
        let sum = 0;
        for (let k = 0; k < this.N; k++) {
          sum += this.get(i, k) * matrix.get(k, j);
        }
        result[calcIndex(i, j, matrix.N)] = sum;
      }
    }
    return new Matrix(result, this.M, matrix.N);
  }

  transpose() {
    const result = new Float64Array(this.M * this.N);
    for (let i = 0; i < this.M; i += 1) {
      for (let j = 0; j < this.N; j += 1) {
        result[calcIndex(j, i, this.M)] = this.get(i, j);
      }
    }
    return new Matrix(result, this.N, this.M);
  }

  omit(col: number) {
    const result = new Float64Array(this.M * this.N - 1);

    let k = 0;
    for (let i = 0; i < this.M; i += 1) {
      for (let j = 0; j < this.N; j += 1) {
        if (col !== j) {
          result[k] = this.get(i, j);
          k += 1;
        }
      }
    }
    return new Matrix(result, this.M, this.N - 1);
  }

  sum(matrix: Matrix) {
    return this.map((x, i, j) => x + matrix.get(i, j));
  }

  hadamard(matrix: Matrix) {
    return this.map((x, i, j) => x * matrix.get(i, j));
  }

  scale(value: number) {
    return this.map((x) => x * value);
  }

  map(mapper: (item: number, i: number, j: number) => number) {
    const result = new Float64Array(this.M * this.N);
    for (let i = 0; i < this.M; i += 1) {
      for (let j = 0; j < this.N; j += 1) {
        result[calcIndex(i, j, this.N)] = mapper(this.get(i, j), i, j);
      }
    }
    return new Matrix(result, this.M, this.N);
  }

  unshift(value: number) {
    if (this.N !== 1) throw new Error("Can only unshift vertical vectors!");
    const result = new Float64Array(this.M * this.N + 1);
    result[0] = value;
    for (let i = 0; i < this.M * this.N; i += 1) {
      result[i + 1] = this.items[i];
    }
    return new Matrix(result, this.M + 1, this.N);
  }

  shift() {
    if (this.N !== 1) throw new Error("Can only shift vertical vectors!");
    const result = new Float64Array(this.M * this.N - 1);
    for (let i = 1; i < this.M * this.N; i += 1) {
      result[i - 1] = this.items[i];
    }
    return new Matrix(result, this.M - 1, this.N);
  }

  iterate(iterator: (value: number, i: number) => void) {
    for (let i = 0; i < this.M; i += 1) {
      iterator(this.get(i, 0), i);
    }
  }
}

function calcIndex(i: number, j: number, N: number) {
  return i * N + j;
}

File: rust/src/matrix.rs
use std::convert::From;

#[derive(Debug)]
pub struct Matrix {
    pub items: Vec<f64>,
    pub rows: usize,
    pub cols: usize,
}

impl Matrix {
    pub fn get(&self, i: usize, j: usize) -> f64 {
        self.items[calc_index(i, j, self.cols)]
    }

    pub fn set(&mut self, i: usize, j: usize, value: f64) {
        self.items[calc_index(i, j, self.cols)] = value
    }

    pub fn multiply(&self, matrix: &Matrix) -> Matrix {
        if self.cols != matrix.rows {
            panic!(
                "`Dimension mismatch. Got {}x{} and {}x{}",
                self.rows, self.cols, matrix.rows, matrix.cols
            )
        }
        let mut result: Vec<f64> = vec![0_f64; self.rows * matrix.cols];

        for i in 0..self.rows {
            for j in 0..matrix.cols {
                let mut sum = 0_f64;
                for k in 0..self.cols {
                    sum += self.get(i, k) * matrix.get(k, j);
                }
                result[calc_index(i, j, matrix.cols)] = sum;
            }
        }

        create_matrix(result, self.rows, matrix.cols)
    }

    pub fn transpose(&self) -> Matrix {
        let mut result: Vec<f64> = vec![0_f64; self.rows * self.cols];
        for i in 0..self.rows {
            for j in 0..self.cols {
                result[calc_index(j, i, self.rows)] = self.get(i, j)
            }
        }
        create_matrix(result, self.cols, self.rows)
    }

    pub fn omit(&self, col: usize) -> Matrix {
        let mut result: Vec<f64> = vec![0_f64; self.rows * (self.cols - 1)];
        let mut k = 0;
        for i in 0..self.rows {
            for j in 0..self.cols {
                if col != j {
                    result[k] = self.get(i, j);
                    k += 1
                }
            }
        }
        create_matrix(result, self.rows, self.cols - 1)
    }

    pub fn sum(&self, matrix: &Matrix) -> Matrix {
        return self.map(&|item: f64, i: usize, j: usize| -> f64 { item + matrix.get(i, j) });
    }

    pub fn subtract(&self, matrix: &Matrix) -> Matrix {
        return self.map(&|item: f64, i: usize, j: usize| -> f64 { item - matrix.get(i, j) });
    }

    pub fn hadamard(&self, matrix: &Matrix) -> Matrix {
        return self.map(&|item: f64, i: usize, j: usize| -> f64 { item * matrix.get(i, j) });
    }

    pub fn scale(&self, value: f64) -> Matrix {
        return self.map(&|_, i: usize, j: usize| -> f64 { value * self.get(i, j) });
    }

    pub fn map(&self, mapper: &dyn Fn(f64, usize, usize) -> f64) -> Matrix {
        let mut result: Vec<f64> = vec![0_f64; self.rows * self.cols];
        for i in 0..self.rows {
            for j in 0..self.cols {
                result[calc_index(i, j, self.cols)] = mapper(self.get(i, j), i, j)
            }
        }
        create_matrix(result, self.rows, self.cols)
    }

    pub fn iterate(&self, iterator: &mut dyn FnMut(f64, usize)) {
        for i in 0..self.rows {
            iterator(self.get(i, 0), i)
        }
    }

    pub fn unshift(&self, value: f64) -> Matrix {
        if self.cols != 1 {
            panic!("Can only unshift vertical vectors!")
        }
        let mut result: Vec<f64> = vec![0_f64; self.rows * (self.cols + 1)];
        result[0] = value;
        for i in 0..self.rows * self.cols {
            result[i + 1] = self.items[i]
        }
        create_matrix(result, self.rows + 1, self.cols)
    }

    pub fn to_arrays(&self) -> Vec<Vec<f64>> {
        let mut result: Vec<Vec<f64>> = vec![vec![0_f64; self.cols]; self.rows];
        for i in 0..self.rows {
            for j in 0..self.cols {
                result[i][j] = self.get(i, j)
            }
        }
        result
    }

    pub fn print(&self) {
        println!("{:?}", self);
    }
}

impl From<&Vec<f64>> for Matrix {
    fn from(items: &Vec<f64>) -> Matrix {
        let rows = items.len();
        let cols = 1;

        let mut result: Vec<f64> = vec![0_f64; rows];
        for i in 0..rows {
            result[i] = items[i]
        }
        create_matrix(result, rows, cols)
    }
}

impl From<&Vec<Vec<f64>>> for Matrix {
    fn from(items: &Vec<Vec<f64>>) -> Matrix {
        let rows = items.len();
        let cols = items[0].len();

        let mut result: Vec<f64> = vec![0_f64; rows * cols];

        for i in 0..rows {
            for j in 0..cols {
                result[i * cols + j] = items[i][j]
            }
        }
        create_matrix(result, rows, cols)
    }
}

fn create_matrix(items: Vec<f64>, rows: usize, cols: usize) -> Matrix {
    Matrix { items, rows, cols }
}

fn calc_index(i: usize, j: usize, cols: usize) -> usize {
    i * cols + j
}
